lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+  bool isOutput;
+};
+
+// Helper to determine if a fully flattened type needs to be flipped.
+static FIRRTLType getPortType(FlatBundleFieldEntry field) {
+  return field.isOutput ? FlipType::get(field.type) : field.type;
+}
+
+// Convert a nested bundle of fields into a flat list of fields.  This is used
+// when working with instances and mems to flatten them.
+static void flattenBundleTypes(FIRRTLType type, StringRef suffixSoFar,
+                               bool isFlipped,
+                               SmallVectorImpl<FlatBundleFieldEntry> &results) {
+  if (auto flip = type.dyn_cast<FlipType>())
+    return flattenBundleTypes(flip.getElementType(), suffixSoFar, !isFlipped,
+                              results);
+
+  // In the base case we record this field.
+  auto bundle = type.dyn_cast<BundleType>();
+  if (!bundle) {
+    results.push_back({type, suffixSoFar.str(), isFlipped});
+    return;
+  }
+
+  SmallString<16> tmpSuffix(suffixSoFar);
+
+  // Otherwise, we have a bundle type.  Break it down.
+  for (auto &elt : bundle.getElements()) {
+    // Construct the suffix to pass down.
+    tmpSuffix.resize(suffixSoFar.size());
+    tmpSuffix.push_back('_');
+    tmpSuffix.append(elt.first.strref());
+    // Recursively process subelements.
+    flattenBundleTypes(elt.second, tmpSuffix, isFlipped, results);
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct FIRRTLTypesLowering
+    : public LowerFIRRTLTypesBase<FIRRTLTypesLowering>,
+      public FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult> {
+
+  using ValueField = std::pair<Value, std::string>;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitDecl;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitExpr;
+  using FIRRTLVisitor<FIRRTLTypesLowering, LogicalResult>::visitStmt;
+
+  void runOnOperation() override;
+
+  // Lowering module block arguments.
+  void lowerArg(BlockArgument arg, BundleType type);
+  void lowerArg(BlockArgument arg, Type type);
+
+  // Lowering operations.
+  LogicalResult visitDecl(InstanceOp op);
+  LogicalResult visitExpr(SubfieldOp op);
+  LogicalResult visitStmt(ConnectOp op);
+
+  // Skip over any other operations.
+  LogicalResult visitUnhandledOp(Operation *op) { return success(); }
+  LogicalResult visitInvalidOp(Operation *op) { return success(); }
+
+  // Helpers to manage state.
+  Value addArgument(Type type, unsigned oldArgNumber,
+                    StringRef nameSuffix = "");
+
+  void setBundleLowering(Value oldValue, StringRef flatField, Value newValue);
+  Optional<Value> getBundleLowering(Value oldValue, StringRef flatField);
+  void getAllBundleLowerings(Value oldValue, SmallVectorImpl<Value> &results);
+
+  void setInstanceLowering(Value oldValue, Value newValue);
+  Value getInstanceLowering(Value oldValue);
```

Comment:
Instead of keeping lowered instances and subfields in different maps, I'd recommend changing this to storing this info in something like:  `DenseMap<std::pair<Value, unsigned>>, Value>`

The key is the old value (an instance, argument, or subfield) and the unsigned is the index into the flattenBundleFields list.  This will make the user-side of a bundle be able to handle it invariant of who produces it, simplifying `visitExpr(SubfieldOp op)` and the input side of instances.

Reply:
I think this makes sense. I will take a look and update the PR.